### PR TITLE
libsamplerate: remove example program unless built with libsndfile

### LINKED ISSUE #47133: libsamplerate build is broken
Hi,

The libsamplerate build is broken because the program sndfile-resample is not working:

```
audio:~ cochlea$ /Users/cochlea/bin//sndfile-resample "-to" "176400" "-c" "0" "/Users/cochlea/test-data//audio-in/mini-drum-loop.wav" "/Users/cochlea/tmp//4/audio-in-0-176400.wav"

****************************************************************
  This example program was compiled without libsndfile 
  (http://www.mega-nerd.com/libsndfile/).
  It is therefore completely broken and non-functional.
****************************************************************

```

### PR DESCRIPTION
**Pull request by rakkesh**
Closes #47133

### CODE REVIEW

File: Library/Formula/libsamplerate.rb [hunk 1]
```diff
@@ -26,6 +26,21 @@ def install
     system "./configure", "--disable-dependency-tracking",
                           "--prefix=#{prefix}"
     system "make", "install"
+    if build.without? "libsndfile"
+      rm_f "#{bin}//sndfile-resample"
```
Comment (DomT4) [inline Library/Formula/libsamplerate.rb]: You can probably stick this on one line:
```ruby
rm_f "#{bin}/sndfile-resample" if build.without? "libsndfile"
```

File: Library/Formula/libsamplerate.rb [hunk 2]
```diff
@@ -26,6 +26,21 @@ def install
...
+
+  def caveats
+    s = ""
+    if build.without? "libsndfile"
+      s += <<-EOS.undent
+      The example program, "sndfile-resample", if not compiled with libsndfile
+      is completely broken and non-functional. (http://www.mega-nerd.com/libsndfile/).
+      Unless "libsndfile" is used as an option, "sndfile-resample" will be removed.
+      EOS
+      s
```
Comment (DomT4) [inline Library/Formula/libsamplerate.rb]: The `s` here should be outside the `end` below. Check the `node` formula if you want an example :)

File: Library/Formula/libsamplerate.rb [hunk 3]
```diff
@@ -26,6 +26,19 @@ def install
     system "./configure", "--disable-dependency-tracking",
                           "--prefix=#{prefix}"
     system "make", "install"
+    rm_f "#{bin}/sndfile-resample" if build.without? "libsndfile"
```
Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: Could you include a comment describing why it's removed, and linking to the GitHub issue for it?

File: Library/Formula/libsamplerate.rb [hunk 4]
```diff
@@ -26,6 +26,19 @@ def install
...
+  end
+
+  def caveats
+    s = ""
```
Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: I think empty caveats should be `nil` instead of the `""` empty string.

File: Library/Formula/libsamplerate.rb [hunk 5]
```diff
@@ -26,6 +26,19 @@ def install
...
+    if build.without? "libsndfile"
+      s += <<-EOS.undent
+      The example program, "sndfile-resample", if not compiled with libsndfile
```
Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: This is a pretty long caveat for a pretty small change. Normally caveats are for big changes in functionality or things the user needs to take action on. And the words "completely broken" are kind of scary. Could you reduce this to a single, milder-sounding line (e.g. "sndfile-resample example is disabled because this was not built with libsndfile"), or remove it entirely?

**Comment by apjanke** [comment]
This seems like a good approach, though maybe it should be less alarmed-sounding, to avoid worrying users. I've put specific comments on the diff.

File: Library/Formula/libsamplerate.rb (hunk 4 above)
Comment (DomT4) [inline Library/Formula/libsamplerate.rb]: To be fair we use `s = ""` in every other formula that sets a blank initial caveat AFAIK.

Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: Oops. Nevermind then; I was just comparing it to the superclass implementation.

Comment (DomT4) [inline Library/Formula/libsamplerate.rb]: If there's no specific reason for the empty string I'm happy for it to be changed, but it's nice to be consistent so we don't end up with a weird mix and contributors aren't sure what to use, so let's do 'em all if a decision is made to rewrite them that way :smile:.

Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: On the one hand, if the empty string is supported by our code, it probably makes for more readable code to use it because then you can just do `s = ""` and have a sequence of conditionalized appends, without having to guard against `nil` in each of them. And if you really want to return `nil`, the nicest idiom might just be to do that anyway and have the last line be `s == "" ? nil : s` to nil-ify it.

On the other hand, client code which calls `caveats` might prefer to have it be `nil` if it's not populated, so they can then just test for `nil`, or use it as a logical value.

I'm no Ruby expert, but using `nil` to represent "no caveats" seems the cleanest way to do it, because then you have a distinct value with a simple test provided by the language's basic types.

Comment (rakkesh) [inline Library/Formula/libsamplerate.rb]: For the time being, I have used the convention followed in other formulae (s = "", instead of s = nil).

**Comment by rakkesh** [comment]
@apjanke updated the PR as per the comments.

File: Library/Formula/libsamplerate.rb (hunk 4 above)
Comment (apjanke) [inline Library/Formula/libsamplerate.rb]: Let's just stick with that then. If we really need to revisit it, we can do so under a separate PR specific to that change.

File: Library/Formula/libsamplerate.rb [hunk 6]
```diff
@@ -33,11 +47,11 @@ def install
 --- a/examples/audio_out.c	2011-07-12 16:57:31.000000000 -0700
 +++ b/examples/audio_out.c	2012-03-11 20:48:57.000000000 -0700
 @@ -168,7 +168,7 @@
- 
+
```
Comment (xu-cheng) [inline Library/Formula/libsamplerate.rb]: Space in patch is crucial. Please don't change them.

File: Library/Formula/libsamplerate.rb [hunk 7]
```diff
@@ -26,6 +26,20 @@ def install
...
+    # https://github.com/Homebrew/homebrew/issues/47133
+    # remove the broken example program from installation
+    rm_f "#{bin}/sndfile-resample" if build.without? "libsndfile"
+  end
+
+  def caveats
+    s = ""
+    if build.without? "libsndfile"
+      s += <<-EOS.undent
+      The example program, "sndfile-resample", is disabled because this formula was not built with libsndfile.
```
Comment (mikemcquaid) [inline Library/Formula/libsamplerate.rb]: Please wrap this line at 80 characters.

**Comment by mikemcquaid** [comment]
Thanks for the PR. One final tweak/nitpick and this should be good to go :+1:

**Comment by rakkesh** [comment]
I'll fix that. By the way, is there a way to write up different caveats depending on the type of brew operation being performed, e.g., `brew info` or `brew install`?

**Comment by mikemcquaid** [comment]
> By the way, is there a way to write up different caveats depending on the type of brew operation being performed, e.g., brew info or brew install?

Nope, they are intentionally the same.

**Comment by DomT4** [comment]
I somewhat brutally hacked up the commit message to tuck it inside the recommended 50 characters, but left the longer one underneath.

Merged in https://github.com/Homebrew/homebrew/commit/4a8654c32d0678cd49923140f7971b4b82caea6d. Thank you for another contribution to Homebrew @rakkesh! :smiley_cat:

**Comment by rakkesh** [comment]
@DomT4 Good job squashing the commit message 👏. I'll try to keep it within the limit next time.
